FLT/Mathlib/Topology/Algebra/Group/Units.lean

```diff
@@ -7,9 +7,88 @@ lemma Submonoid.units_isOpen {M : Type*} [TopologicalSpace M] [Monoid M]
   {U : Submonoid M} (hU : IsOpen (U : Set M)) : IsOpen (U.units : Set Mˣ) :=
   (hU.preimage Units.continuous_val).inter (hU.preimage Units.continuous_coe_inv)
 
+ -- FLT#588
 lemma Submonoid.units_isCompact {M : Type*} [TopologicalSpace M] [Monoid M] [ContinuousMul M]
     [T2Space M] {U : Submonoid M} (hU : IsCompact (U : Set M)) : IsCompact (U.units : Set Mˣ) := by
-  sorry -- FLT#588
+  let φ : Mˣ → M × M := fun x ↦  (↑x, ↑x⁻¹)
+  let p : M × M → M := fun (u,v) => u * v
+  let q : M × M → M := fun (u,v) => v * u
+
+  have range_φ : p ⁻¹' {1} ∩ q ⁻¹' {1} = Set.range φ := by
+    apply Set.ext
+    intro (u,v)
+    simp_all only[p,q,φ]
+    simp_all only [Set.image_univ, Set.mem_range, Prod.mk.injEq,
+         Set.mem_inter_iff, Set.mem_preimage, Set.mem_singleton_iff]
+    constructor
+    · -- Forward direction: u * v = 1 ∧ v * u = 1 ⇒ (u, v) ∈ range φ
+      intro h
+      let y : Mˣ := ⟨u, v, h.1, h.2⟩
+      use y
+      simp only [Units.inv_mk, and_self, y]
+    · -- Backward direction: (u, v) ∈ Set.range φ ⇒ (u * v = 1 ∧ v * u = 1)
+      intro h
+      obtain ⟨w, h⟩ := h
+      obtain ⟨left, right⟩ := h
+      subst left right
+      simp_all only [Units.mul_inv, Units.inv_mul, and_self, q, φ, p]
```

Comment:
```suggestion
  have range_φ : p ⁻¹' {1} ∩ q ⁻¹' {1} = Set.range φ := by
    ext ⟨u,v⟩ -- better to use a tactic than having to know the name of the extensionality lemma
    constructor
    · -- Forward direction: u * v = 1 ∧ v * u = 1 ⇒ (u, v) ∈ range φ
      intro h
      use ⟨u, v, h.1, h.2⟩
      rfl -- true by definition (definitional equality is important in dependent type theory)
    · -- Backward direction: (u, v) ∈ Set.range φ ⇒ (u * v = 1 ∧ v * u = 1)
      rintro ⟨w, rfl, rfl⟩ -- does your intro, obtain, subst all in one go
      simp [p, q] -- simplifier can deal with it now.
```
Note that my instinct is to avoid all names of lemmas if I can.